docs/smart-contracts/meebits_erc721.sol

```diff
@@ -0,0 +1,679 @@
+
+    function _safeTransferFrom(address _from,  address _to,  uint256 _tokenId,  bytes memory _data) private canTransfer(_tokenId) validNFToken(_tokenId) {
+        address tokenOwner = idToOwner[_tokenId];
+        require(tokenOwner == _from, "Incorrect owner.");
+        require(_to != address(0));
+
+        _transfer(_to, _tokenId);
+
+        if (isContract(_to)) {
+            bytes4 retval = ERC721TokenReceiver(_to).onERC721Received(msg.sender, _from, _tokenId, _data);
+            require(retval == MAGIC_ON_ERC721_RECEIVED);
+        }
+    }
+
+    function _clearApproval(uint256 _tokenId) private {
+        if (idToApproval[_tokenId] != address(0)) {
+            delete idToApproval[_tokenId];
+        }
+    }
+
+    //// Enumerable
+
+    function totalSupply() public view returns (uint256) {
+        return numTokens;
+    }
+
+    function tokenByIndex(uint256 index) public pure returns (uint256) {
+        require(index >= 0 && index < TOKEN_LIMIT);
+        return index + 1;
+    }
+
+    function tokenOfOwnerByIndex(address _owner, uint256 _index) external view returns (uint256) {
+        require(_index < ownerToIds[_owner].length);
+        return ownerToIds[_owner][_index];
+    }
+
+    //// Metadata
+
+    /**
+      * @dev Converts a `uint256` to its ASCII `string` representation.
+      */
+    function toString(uint256 value) internal pure returns (string memory) {
+        if (value == 0) {
+            return "0";
+        }
+        uint256 temp = value;
+        uint256 digits;
+        while (temp != 0) {
+            digits++;
+            temp /= 10;
+        }
+        bytes memory buffer = new bytes(digits);
+        uint256 index = digits - 1;
+        temp = value;
+        while (temp != 0) {
+            buffer[index--] = bytes1(uint8(48 + temp % 10));
+            temp /= 10;
+        }
+        return string(buffer);
+    }
+
+    /**
+      * @dev Returns a descriptive name for a collection of NFTokens.
+      * @return _name Representing name.
+      */
+    function name() external view returns (string memory _name) {
+        _name = nftName;
+    }
+
+    /**
+     * @dev Returns an abbreviated name for NFTokens.
+     * @return _symbol Representing symbol.
+     */
+    function symbol() external view returns (string memory _symbol) {
+        _symbol = nftSymbol;
+    }
+
+    /**
+     * @dev A distinct URI (RFC 3986) for a given NFT.
+     * @param _tokenId Id for which we want uri.
+     * @return _tokenId URI of _tokenId.
+     */
+    function tokenURI(uint256 _tokenId) external view validNFToken(_tokenId) returns (string memory) {
+        return string(abi.encodePacked("https://meebits.larvalabs.com/meebit/", toString(_tokenId)));
+    }
+
+    //// MARKET
+
+    struct Offer {
+        address maker;
+        address taker;
+        uint256 makerWei;
+        uint256[] makerIds;
+        uint256 takerWei;
+        uint256[] takerIds;
+        uint256 expiry;
+        uint256 salt;
+    }
```

Comment:
Represent an Offer created on the developer site between a maker and a taker. 
Can be used to trade meebits with additional ether (as ether fraction wei) on either sides